flink-python/dev/lint-python.sh

```diff
@@ -0,0 +1,508 @@
+
+    print_function "STAGE" "Stage 3:flake8 codestyle check"
+    if [ ! -f "$FLAKE8_PATH" ]; then
+        echo "For some unkown reasons, the flake8 package is not complete,\
+        you should exec the script with the parameter: -f"
+    fi
+
+    if [[ ! "$PYTHON_SOURCE" ]]; then
+        echo "No python files found!  Something is wrong exiting."
+        exit 1;
+    fi
+
+    ($FLAKE8_PATH  --config=tox.ini $PYTHON_SOURCE) 2>&1 | tee -a $LOG_FILE
+
+    PYCODESTYLE_STATUS=$?
+    if [ $PYCODESTYLE_STATUS -ne 0 ]; then
+        print_function "STAGE" "pycodestyle checks ... [FAILED]"
+    else
+        print_function "STAGE" "pycodestyle checks ... [SUCCESS]"
+    fi
+}
+
+# CURRENT_DIR is "flink/flink-python/dev/"
+CURRENT_DIR="$(cd "$( dirname "$0" )" && pwd)"
+
+# FLINK_PYTHON_DIR is "flink/flink-python"
+FLINK_PYTHON_DIR=$(dirname "$CURRENT_DIR")
+pushd "$FLINK_PYTHON_DIR" &> /dev/null
+
+# conda path
+CONDA_PATH=$CURRENT_DIR/.conda/bin/conda
+
+# tox path
+TOX_PATH=$CURRENT_DIR/.conda/bin/tox
+
+# flake8 path
+FLAKE8_PATH=$CURRENT_DIR/.conda/bin/flake8
+
+_OLD_PATH="$PATH"
+
+SUPPORT_OS=("Darwin" "Linux")
+
+# the file stores the success progress.
+STAGE_FILE=$CURRENT_DIR/.stage.txt
+
+# the dir includes all kinds of py env installed.
+VIRTUAL_ENV=$CURRENT_DIR/.conda/envs
+
+LOG_DIR=$CURRENT_DIR/log
+
+if [ "$FLINK_IDENT_STRING" == "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+if [ "$HOSTNAME" == "" ]; then
+    HOSTNAME="$HOST"
+fi
+
+# the log file stores the checking result.
+LOG_FILE=$LOG_DIR/flink-$FLINK_IDENT_STRING-python-$HOSTNAME.log
+create_dir $LOG_DIR
+if [ $? -ne 0 ]; then
+    echo "create file failed."
+    exit 1
+fi
+# clean LOG_FILE content
+echo >$LOG_FILE
+
+# miniconda script
+CONDA_INSTALL_SH=$CURRENT_DIR/download/miniconda.sh
+
+# stage "install" includes the num of steps.
+STAGE_INSTALL_STEPS=5
+
+# whether force to restart the script.
+FORCE_START=0
+# parse_opts
+USAGE="
+usage: $0 [options]$
+-h           print this help message and exit
+-f           force to exec from the start
+"
+while getopts "hf" arg; do
+    case "$arg" in
+        h)
+            printf "%s\\n" "$USAGE"
+            exit 2
+            ;;
+        f)
+            FORCE_START=1
+            ;;
+        ?)
+            printf "ERROR: did not recognize option '%s', please try -h\\n" "$1"
+            exit 1
+            ;;
+    esac
+done
+# If exec the script with the param: -f, all progress will be re-run
+if [ $FORCE_START -eq 1 ]; then
+    STAGE="install"
+    STEP=0
+    checkpoint_stage $STAGE $STEP
+else
+    restore_stage
+fi
+# Stage 1:install environment
+if [ $STAGE == "install" ]; then
+    install_environment
+fi
+# Stage 2:tox check python compatibility
+if [ $STAGE == "tox" ]; then
+    tox_check
+fi
+# Stage 3:code style test
+if [ $STAGE == "flake8" ]; then
+    flake8_check
+fi
+
+echo "All the checks are finished, the detailed information can be found in: $LOG_FILE"
```

Comment:
IMO we should exit with a non-zero code when any checks failed just like we did in FLINK_HOME/tools/travis_controller.sh. The reason is that Travis-CI determines whether the script is executed successfully according to whether the exit code is 0 or not.

Reply:
Yes, I think we should quit immediately when we encounter an error.